src/aap_eda/api/serializers/activation.py

```diff
@@ -473,13 +471,188 @@ def create(self, validated_data):
+        allow_null=True,
+        allow_blank=True,
+        validators=[validators.check_if_rfc_1035_compliant],
+    )
+
+    def validate(self, data):
+        _validate_credentials_and_token_and_rulebook(data=data, creating=True)
+        _validate_sources_with_event_streams(data=data)
+        return data
+
+    def prepare_update(self, activation: models.Activation):
+        rulebook_id = self.validated_data.get("rulebook_id")
+        self.validated_data["user_id"] = self.context["request"].user.id
+        if rulebook_id:
+            rulebook = models.Rulebook.objects.get(id=rulebook_id)
+            self.validated_data["rulebook_name"] = rulebook.name
+            self.validated_data["rulebook_rulesets"] = rulebook.rulesets
+            self.validated_data["git_hash"] = rulebook.project.git_hash
+            self.validated_data["project_id"] = rulebook.project.id
+
+        system_vault_credential = activation.eda_system_vault_credential
+        if system_vault_credential:
+            inputs = yaml.safe_load(
+                system_vault_credential.inputs.get_secret_value()
+            )
+            vault_data = VaultData(inputs["vault_password"], True)
+        else:
+            vault_data = VaultData()
+        if self.validated_data.get("source_mappings", []):
+            if not rulebook_id:
+                # load the original ruleset
+                self.validated_data[
+                    "rulebook_rulesets"
+                ] = activation.rulebook.rulesets
+
+            # update with new event streams
+            self.validated_data[
+                "rulebook_rulesets"
+            ] = _update_event_stream_source(self.validated_data, vault_data)
+        elif rulebook_id:
+            # user selects a rulebook but no mapping, clear existing mappings
+            self.validated_data["source_mappings"] = ""
+            self.validated_data["event_streams"] = []
+
+        if self.validated_data.get("eda_credentials"):
+            _update_extra_vars_from_eda_credentials(
+                validated_data=self.validated_data,
+                vault_data=vault_data,
+                creating=True,
+            )
+
+        if vault_data.password_used and not system_vault_credential:
+            self.validated_data[
+                "eda_system_vault_credential"
+            ] = _create_system_eda_credential(
+                vault_data.password,
+                _get_vault_credential_type(),
+                self.validated_data.get(
+                    "organization_id", activation.organization.id
+                ),
+            )
+
+    def update(
+        self, instance: models.Activation, validated_data: dict
+    ) -> None:
+        update_fields = []
+        eda_credentials = None
+        event_streams = None
+        for key, value in validated_data.items():
+            if key == "eda_credentials":
+                eda_credentials = value
+                continue
+            elif key == "event_streams":
```

Comment:
@bzwei The event_streams are built from the source mappings since we have to support multiple sources in a rulebook, the rulebook could have multiple sources and the source mappings tells us which which source to replace with which event stream, you could have sources that have to kept intact, so we can't be directly accepting the event_streams from the caller. The event_streams should get cleared based on the presence of source_mappings and regenerated.

Reply:
The event source mappings for updating need to be completed. event_streams here are rebuilt from the source mappings given by user.